### ISSUE TITLE
Possibly redundant OpenSSL empty fragment workaround code

### ISSUE THREAD

**Issue by i-rinat** [ARMmbed/mbedtls#2883]
### Description
- Type: Question
- Priority: Minor

## Question
What the code below is for, exactly?
There is a comment about OpenSSL sending empty messages, but in the memo (https://www.openssl.org/~bodo/tls-cbc.txt) they say about empty plaintext fragments, which are transformed to non-empty ciphertexts. The trick changed a current IV, and made sense as a workaround. Sending empty ciphertext doesn't look productive at all. No internal state is changed, so it cannot affect ciphering/deciphering process.

https://github.com/ARMmbed/mbedtls/blob/74692aeb8c7e2c1c0f273592d87472a6f2aabacf/library/ssl_tls.c#L10766-L10780

I strongly believe this code was written in false understanding of the memo, and is better to be removed. If any OpenSSL version with enabled workaround sends an empty plaintext, the ciphertext should be successfully deciphered to an empty plaintext, and thus will not affect the application at all. That was the purpose of such workaround — to disrupt applications as less as possible, so that the existing ones will continue to work without any need of implementing a special case.

I would like to hear any comments on the issue. Also, if the code is actually redundant and to be removed, could anyone with already signed CLA apply the changes?

**Comment by gilles-peskine-arm** [comment]
I agree that this comment doesn't look right. But I wouldn't want to remove code without understanding what it's for.

Tracking the origin of this code, I see that it was added between XySSL 0.7 and XySSL 0.8. Unfortunately, I don't have a detailed history from the days before PolarSSL. I see the following entry in `BUGS.txt` in XySSL 0.8:

> 07/10/2007  goldenchild: (bug) OpenSSL sends empty records

That's under `[FORUM]`. This forum is long gone and I can't find 2007 posts on the Wayback Machine. Maybe Arm has inherited an archive of this forum somewhere, I'll inquire.

**Comment by hanno-arm** [comment]
The code looks correct to me: `mbedtls_ssl_read_record()` includes record decryption, so the check in question applies to the _plaintext length_, not the ciphertext length. The code then immediately attempts to fetch another record when finding an empty AD record, which seems to be a valid shortcut -- albeit not a necessary one, since the subsequent logic handling incoming application records appears to handle the case of length `0` records correctly.

I don't think a change is needed here.

**Comment by mpg** [comment]
I tend to agree with Hanno, though maybe we should clarify the comment in order to avoid any future misunderstandings.

**Comment by hanno-arm** [comment]
I wouldn't object to a patch clarifying the comment, but given that we intend to replace the messaging layer in the (near?) future anyway, I'd rather leave it for now / place it low on the priority list.

**Comment by gilles-peskine-arm** [comment]
I'd prefer to clarify the comment. It would take less time than to discuss it.
